### PR TITLE
PT-2670: Improve PatientData interface

### PR DESCRIPTION
**Pull request by allasm**
@itaiGershtansky @veronikaslc please review

### CODE REVIEW

File: components/patient-data/api/src/main/java/org/phenotips/data/PatientData.java [hunk 1]
```diff
@@ -100,6 +100,15 @@
     boolean isNamed();
 
     /**
+     * For dictionary data only, checks if the given key is present in the dictionary data.
+     *
+     * @param key the key to be checked for
+     * @return {@code true} if the key-value based data structure has the specified key
+     * @since 1.0M13
```
Comment (veronikaslc) [inline components/patient-data/api/src/main/java/org/phenotips/data/PatientData.java]: @since 1.3M2?

Comment (veronikaslc) [inline components/patient-data/api/src/main/java/org/phenotips/data/PatientData.java]: +1 other than that minor stuff

**Comment by veronikaslc** [comment]
+1

**Comment by itaiGershtansky** [comment]
+1

**Comment by sdumitriu** [comment]
`Map` uses `containsKey` instead of `hasKey`. And while we're at it, why not also add `containsValue`?

**Comment by allasm** [comment]
containsKey() probably makes sense for consistency - will amend

...but I don't see a use for containsValue() in the context of PatientData, as of right now (when the need arises we'll add it?). That, and the fact that containsKey() is probably more efficient when implemented using equivalent internal map method as opposed to using keyIterator, while containsValue is implemented with the same efficiency internally as it can be using existing valueIterator.

**Comment by sdumitriu** [comment]
Not really, `contains` also uses hashcode and equals, so for a large map it will be faster.

What's different is that I can't find a usecase for it.

**Comment by allasm** [comment]
I updated the method name.

I checked a few implementations of `containsVallue()` in standard maps in java.util, and all of them seem to be O(n) or O(hash size) or alike, which is generally comparable to O(n)... given that and the current lack of use for `constainsValue()`, we should probably not clutter the interface with an extra method, which only makes sense for Dictionary implementation of `PatientData` to start with.

By the way, I think we over-engineered PatientData, and we should only have one implementation, the dictionary one, and `Indexed` should use dictionary with keys being numbers, and `Simple` use dict. with one controller-defined key.  (possibly an extension of Dictionary implementation with a convenient constructor)

**Pull request merged**